llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -6905,6 +6905,17 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
       if (isa<VPPartialReductionRecipe>(&R))
         return true;
 
+      // The VPlan-based cost model can analyze if recipes are scalar
+      // recursively, but legacy cost model cannot.
+      if (auto *WidenMemR = dyn_cast<VPWidenMemoryRecipe>(&R)) {
+        auto *AddrI = dyn_cast<Instruction>(
+            getLoadStorePointerOperand(&WidenMemR->getIngredient()));
+        if (WidenMemR && AddrI &&
+            vputils::isSingleScalar(WidenMemR->getAddr()) !=
+                CostCtx.isLegacyUniformAfterVectorization(AddrI, VF))
+          return true;
+      }
```

Comment:
The WidenMemR check is redundant 
```suggestion
      if (auto *WidenMemR = dyn_cast<VPWidenMemoryRecipe>(&R))
        if (auto *AddrI = dyn_cast<Instruction>(
            getLoadStorePointerOperand(&WidenMemR->getIngredient()));
          if (vputils::isSingleScalar(WidenMemR->getAddr()) !=
                CostCtx.isLegacyUniformAfterVectorization(AddrI, VF))
            return true;
```